fdbclient/SpecialKeySpace.actor.cpp

```diff
@@ -980,69 +958,117 @@ ACTOR Future<bool> checkExclusion(Database db,
+	state std::map<std::string, std::vector<std::string>> parsedLocalities;
+	// Convert the passed localities into a map of vectors to make it easier to check if a process
+	// is excluded by locality.
+	for (const auto& locality : *localities) {
+		std::pair<std::string, std::string> locality_key_value = decodeLocality(locality);
+		if (locality_key_value.first == "") {
+			continue;
+		}
+
+		parsedLocalities[locality_key_value.first].push_back(locality_key_value.second);
+	}
 
 	try {
 		for (auto proc : processesMap.obj()) {
 			StatusObjectReader process(proc.second);
-			std::string addrStr;
-			if (!process.get("address", addrStr)) {
-				*msg = ManagementAPIError::toJsonString(false, markFailed ? "exclude failed" : "exclude", errorString);
-				return false;
+			StatusObjectReader localityObj;
+			bool hasLocalities = process.get("locality", localityObj);
+
+			bool excluded = process.has("excluded") && process.last().get_bool();
+			// If the process is already excluded based on the process status, we don't have to check if the process is
+			// also excluded by address or locality.
+			if (!excluded) {
+				// First check if the process is excluded based on the exclusions.
+				std::string addrStr;
+				if (!process.get("address", addrStr)) {
+					*msg =
+					    ManagementAPIError::toJsonString(false, markFailed ? "exclude failed" : "exclude", errorString);
+					return false;
+				}
+				NetworkAddress addr = NetworkAddress::parse(addrStr);
+				excluded = addressExcluded(*exclusions, addr);
+
+				// If the process is not already excluded and the parsed localites have at least one entry,
+				// check if the process is excluded by localities.
+				if (!excluded && hasLocalities && !parsedLocalities.empty()) {
+					// Iterate over all excluded localites and check if the process is excluded based on the locality.
+					for (std::map<std::string, std::vector<std::string>>::iterator it = parsedLocalities.begin();
+					     it != parsedLocalities.end();
+					     ++it) {
+						std::string localityValue;
+						if (!localityObj.get(it->first, localityValue)) {
+							// If the locality doesn't exist in the locality object skip over it.
+							continue;
+						}
+
+						std::vector<std::string> localityVec = it->second;
+						// When the process has a matching locality field, add it to the exclusion list.
+						if (std::find(localityVec.begin(), localityVec.end(), localityValue) != localityVec.end()) {
+							excluded = true;
+							// If the exclusion will exclude the processes as failed, we have to store the addresses of
+							// the excluded processes in this list, as the checkSafeExclusions check will work on this
+							// information.
+							if (markFailed) {
+								auto addrExclusion = AddressExclusion(addr.ip, addr.port);
+								if (std::find(addresses->begin(), addresses->end(), addrExclusion) !=
+								    addresses->end()) {
+									addresses->push_back(addrExclusion);
```

Comment:
this is surprising--modifying the input parameter. Can you add a comment for this actor, documenting this behavior?

Reply:
I added a comment above the `if (markFailed) {` statement. If we don't want to modify the input, I can create a new vector that contains all the addresses in case of a `failed` exclusion.

Reply:
I decided to update the code to not modify the input vector.